### PR TITLE
[AdvancedSettings] change defaults for watched state and resume point

### PR DESCRIPTION
**Pull request by DaVukovic**
## Description

This changes the default behavior of Kodi if users import libraries from NFO files to import watched states and resume points by default. 

As this is not a fix but a improvement, I won't do a backport of that.

## Motivation and context

If users already do have NFO files, watched states and resume points are already written to those files. Why should we not use this information by default and force users to create and edit files on systems on which those files are complicated to create (Android, XBOX, LibreELEC). So I guess this is very helpful for users. 

If users don't want to have watched states to be imported at all, they can open the context menu from the "Titles"-section and mark the full library as "unwatched":

image

It still can be disabled via the advancedsettings.xml. Existing advancedsettings.xml files where this is already set to "true" shouldn't be affected as they would simply overwrite the default setting with the same value. 

## What is the effect on users?

Users aren't forced anymore to create files on systems they don't fully understand or are not familiar with.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by KarellenX** [comment]
Thanks @DaVukovic for this change.

I agree with the changes. 

We have had a couple of settings which hide library items or omit library data by default and forces a user to change GUI settings or make as.xml entries which are difficult for some users. (The other was Hide empty TV Shows). For many users, these requirements are not known to them which causes them some level of confusion when they get unexpected results.

Allowing Play Count, Watched Status and Resume Points to be imported by default will reduce confusion for many users. A typical example is in this thread today... https://forum.kodi.tv/showthread.php?tid=366445&pid=3080335#pid3080335

Presently, there is only one way to scan this data, a user is forced to make an as.xml entry.

If we reverse the setting and enable it by default then a user has two ways to remove this data
1. By making an as.xml entry with values of False
2. By selecting Mark Unwatched from the context menu which can selectively clear this data on a per show/movie basis, or the entire movie/tv show library.

Yes there is a Mark Watched selection also, but this sets the current date for the Last Played date. I don't think users want their entire library as Watched today. Also it sets a play count of 1, and favourite movies/episodes may have been watched more than once.

**Comment by enen92** [comment]
There's some background of why this is by default false in the PR that introduced one of the settings (git blame): https://github.com/xbmc/xbmc/pull/895

I can relate to some of the reasoning but don't have strong feelings here. Furthermore it seems that both settings are bypassed (and the info is imported into the database) if doing a full library import. Won't this be more an issue of lack of documentation that the actual functionality?

**Comment by DaVukovic** [comment]
I've read that PR before I've create this one. And tbh, some of those comments do make some sense. But those comments are about 10 years old and nothing changed in between. We still use NFO files to store such information. We still don't do SQL dumps to export the database. 
What has changed in between are the systems Kodi is running on and the users who are using Kodi. Therefore the creation of advancedsettings.xml files gets more complicated for users who simply want to use Kodi.

The comment that this is only needed on your very first installation of Kodi is something I don't agree on. I guess nearly everyone will agree that at some point we all have re-initialized our database. Maybe it be for testing purposes or may it be because the public-joe broke his library. I've heard things like "if I have to delete and reinstall I'll loose all my watched states" way to often and explained it many times to users and also have dealt enough to explain how to copy and create files on different systems. 

The only thing I agree on is that we should do real SQL dumps instead of exporting to NFO files. But I guess as that comment is 10 years old and nothing changed in between, that we can now say that this won't happen in the near future ;). So what harm does it do to simply make Kodi easier for users for the time being and change that again in case we will ever do real SQL dumps?

>I'm against exporting and importing resume points it could be months later that you need tip reimport a file or maybe even a different PC and the information would be way way way out of date and your stuck with it unless you do the overwrite old files on export

This for example....yeah, the info might be out of date. But I guess that info then is only lacking items I have additionally watched. Movies I have watched will most likely not be marked as unwatched anymore. So wouldn't it be good to have at least an outdated info rather than no info at all? 

If imported entries on the DB are marked as watched and if I don't want that, then it's easy to mark them all as unwatched (see above). Unfortunately Estuary doesn't provide a setting to not show resume or watched icons (or I missed it). So if I watch something, I'll get them anyway and I somewhat doubt that users mark movies or tv-shows as "unwatched" after watching them.

Probably I'm on the wrong side, but I would say we should have a look on what the user currently needs and not what developers would have in mind what might be good but never gets implemented ;) (for sure no offense to anyone).

File: xbmc/settings/AdvancedSettings.cpp
Comment (enen92) [inline xbmc/settings/AdvancedSettings.cpp:292]:
```suggestion
  m_bVideoLibraryImportWatchedState{true};
```

See: https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#114-default-member-initialization

File: xbmc/settings/AdvancedSettings.cpp
Comment (enen92) [inline xbmc/settings/AdvancedSettings.cpp:293]:
```suggestion
  m_bVideoLibraryImportResumePoint{true};
```

See: https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#114-default-member-initialization

**Review by enen92 — COMMENTED** [review]
As I said, I have no strong feelings here (just shared the PR in case you have missed it). I trust your judgement (and Karelen) to decide if this is wanted by users or not since you're pretty much the most active guys supporting users on the forums. If this is what the majority of people want, Kodi should provide reasonable and sane defaults.

So +1 from me (will approve later after the changes are adjusted)

**Comment by KarellenX** [comment]
Thanks @enen92 

From my experience on the forum, this catches a lot of users out. Its a step they need to make without realising they need to make it. They trust that if they export nfo files, then scanning them back will give them the complete metadata, but its not the case.

I am not even sure why it is a setting. We don't have settings to enable or disable the import of, say, User Rating, or MPAA certifications or the host of other metadata. So why do these particular items require a setting? Especially when you can easily clear that data from within the GUI.

No disrespect intended to the original developers is intended. I am assuming that use case has changed in 10 years. And I think that most users that are rebuilding their library, or adding a new setup in their home, should/will export a fresh batch of nfo files. Its hard to imagine users will intentionally use nfo files that are 6 months or a couple of years old, except in the case of a c:\ drive failure.

**Review by enen92 — APPROVED** [review]

File: xbmc/settings/AdvancedSettings.cpp
Comment (DaVukovic) [reply xbmc/settings/AdvancedSettings.cpp]: done

File: xbmc/settings/AdvancedSettings.cpp
Comment (DaVukovic) [reply xbmc/settings/AdvancedSettings.cpp]: done

**Comment by jjd-uk** [comment]
Perhaps no one on that original PR realised it was as simply as do the `Mark as unwatched` on the `Title` node if the data was stale so wasn't of use anymore.

File: xbmc/settings/AdvancedSettings.cpp
Comment (mkreisl) [inline xbmc/settings/AdvancedSettings.cpp:301]: But my gcc 10 does not like that at all

File: xbmc/settings/AdvancedSettings.cpp
Comment (mkreisl) [inline xbmc/settings/AdvancedSettings.cpp:302]: Same ^^

File: xbmc/settings/AdvancedSettings.cpp
Comment (howie-f) [reply xbmc/settings/AdvancedSettings.cpp:301]: shouldnt be a problem, whats the error?

Comment (enen92) [reply xbmc/settings/AdvancedSettings.cpp:301]: ofc...because this is wrong (thought this was the header file).
Sorry @DaVukovic. Please keep it as you had it before

```suggestion
  m_bVideoLibraryImportWatchedState = true;
  m_bVideoLibraryImportResumePoint = true;
```

You could also simply drop this two lines and initialize them to true in the header file (in AvancedSettings.h):

```
bool m_bVideoLibraryImportWatchedState{true};
bool m_bVideoLibraryImportResumePoint{true};
```

Comment (mkreisl) [reply xbmc/settings/AdvancedSettings.cpp:301]:
```
/home/xbian/xbian-package-xbmc/build/rpi4-nexus/working/xbmc/settings/AdvancedSettings.cpp: In member function 'void CAdvancedSettings::Initialize()':
/home/xbian/xbian-package-xbmc/build/rpi4-nexus/working/xbmc/settings/AdvancedSettings.cpp:310:36: error: expected ';' before '{' token
  310 |   m_bVideoLibraryImportWatchedState{true};
      |                                    ^
      |                                    ;
/home/xbian/xbian-package-xbmc/build/rpi4-nexus/working/xbmc/settings/AdvancedSettings.cpp:310:3: warning: statement has no effect [-Wunused-value]
  310 |   m_bVideoLibraryImportWatchedState{true};
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/xbian/xbian-package-xbmc/build/rpi4-nexus/working/xbmc/settings/AdvancedSettings.cpp:311:35: error: expected ';' before '{' token
  311 |   m_bVideoLibraryImportResumePoint{true};
      |                                   ^
      |                                   ;
/home/xbian/xbian-package-xbmc/build/rpi4-nexus/working/xbmc/settings/AdvancedSettings.cpp:311:3: warning: statement has no effect [-Wunused-value]
  311 |   m_bVideoLibraryImportResumePoint{true};
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [build/settings/CMakeFiles/settings.dir/build.make:82: build/settings/CMakeFiles/settings.dir/AdvancedSettings.cpp.o] Error 1
make[2]: Leaving directory '/home/xbian/xbian-package-xbmc/build/rpi4-nexus/working/build'
make[1]: *** [CMakeFiles/Makefile2:7954: build/settings/CMakeFiles/settings.dir/all] Error 2
make[1]: Leaving directory '/home/xbian/xbian-package-xbmc/build/rpi4-nexus/working/build'
make: *** [Makefile:149: all] Error 2
```

Comment (mkreisl) [reply xbmc/settings/AdvancedSettings.cpp:301]: it IS a problem!

Comment (howie-f) [reply xbmc/settings/AdvancedSettings.cpp:301]: whoops

Comment (mkreisl) [reply xbmc/settings/AdvancedSettings.cpp:301]: you should have a look at your outputs of the check build

Comment (lrusak) [reply xbmc/settings/AdvancedSettings.cpp:301]: > you should have a look at your outputs of the check build
> 
> 

It was an honest mistake. It was also caught by our CI.

**Comment by DaVukovic** [comment]
Sorry for the hassle. Did it on the wrong section. I now removed those lines and initialized them with `true` at the header file

**Review by KarellenX — APPROVED** [review]

**Comment by DaVukovic** [comment]
@jjd-uk Sorry, haven't seen your question

Yeah, I did some tests and marking a full library (900+ entries) as watched/unwatched is pretty much instant with a local library on my x86 machine. It might take a bit of time in case of a MySQL DB. But that will also depend on the network in use (LAN/WiFi).

**Comment by jjd-uk** [comment]
> @jjd-uk Sorry, haven't seen your question
> 
> Yeah, I did some tests and marking a full library (900+ entries) as watched/unwatched is pretty much instant with a local library on my x86 machine. It might take a bit of time in case of a MySQL DB. But that will also depend on the network in use (LAN/WiFi).

That seems fine then. I just wanted to be sure doing the `Mark as unwatched` on the` Title` node would not create signifcant delay if we're going to be saying that's the method for blowing out stale data,. Instananous on a 900 titile library probably means it likely to be no more than an order of some seconds even for extremely large libraries, which is fine in my view.

**Comment by KarellenX** [comment]
FWIW, another user struggling with these tags, not realising they are needed...
https://forum.kodi.tv/showthread.php?tid=366445&pid=3081572#pid3081572

**Comment by Klojum** [comment]
@DaVukovic 

**If users already do have NFO files, watched states and resume points are already written to those files.**
Unless I missed something, the default setting for exporting the video library is still without watched states and resume points. Users need to enable it via the advancedsettings.xml file. Can we reverse that, and enable it by default, and that users can disable it whenever necessary?

**We still don't do SQL dumps to export the database.**
MySQLdump could be perhaps included, or mimic'd via our own code.

**The only thing I agree on is that we should do real SQL dumps instead of exporting to NFO files.**
I would disagree on that. Not everyone is SQL-savvy or comfortable with it I think plenty of users will use nfo text files so that they can easily change/fix whatever is necessary. Having an internal mysql dump option as a additional means would be a nice to have. I did a mysqldump Linux script on the Wiki that works pretty well. Creating the backup sql file only takes 2-3 seconds on a medium sized video library, a single second for a small music library. This is on a 'server' PC with the database on a SSD and the export file sent via gigabit network. The video dump file is approx 47 MB in size.

**Comment by jjd-uk** [comment]
Regarding 

**The only thing I agree on is that we should do real SQL dumps instead of exporting to NFO files**

My underrstanding is that NFO's should only contain the metadata picked up from scraping site to save having to go back to them when rescraping items.  So any other data such as watched status, media stream details etc that Kodi stores in the database should not be added to NFO's on export. Instead it would be more appropiate for this additional data to be part of a Backup feature that could be kicked off from the GUI that would contain a database dump.

Peronally I'd agree putting things like Watched status into NFO's is a hack and ideally should be avoided. However it seems no one is going to take on the work to do a proper Backup feature, unless it could be a possible GSOC project, so it seems we've no choice but stick with these hacks.

**Comment by DaVukovic** [comment]
>Unless I missed something, the default setting for exporting the video library is still without watched states and resume points. Users need to enable it via the advancedsettings.xml file. Can we reverse that, and enable it by default, and that users can disable it whenever necessary?

AFAIK there isn't a settings which I have to enable to export watched-states and/or resume points. Those are exported by default if you export your database to single files. That's also what our wiki says about exporting the library:

https://kodi.wiki/view/Import-export_library/Video

Regarding SQL dumps....

For me NFO files should be used to add items to the library which can't be scanned by scrapers (for whatever reason..may it be homemade videos, or something I would have in another way scrapers do provide). From my POV those NFO files are config files how I want Kodi to present my items added to the library. 

All those infos, after scanning them, are stored in a SQL-DB. Dumps of those DBs should then contain the full state of the current library (including watched-states, play counts, resume points, IMDB ratings and all that). I totally agree with the comments from 10 years ago that this would be a real DB-export. I also agree that not everyone is able to do that using additional tools. Hence I would suggest that those dumps should be done via the Kodi GUI. Either by some team-made add-on (which might be hard to create as there doesn't seem to exist a proper way to dump a SQL DB via python) or via a solution directly implemented in Kodi core. Importing those dumps (which would also include importing all the mentioned states) should then be done via the Kodi GUI as well. That would mean, that I don't have to take care much anymore about my NFO files. They won't become useless, but they will be less important if I have a SQL dump. 

Currently the only way to "dump" (which is the totally wrong wording for that) those DBs is to copy the full DB to another location. And that's not even recommended. Anyway, SQL dumps Is not what this PR addressed. 

By using dumps, and if anyone considers to implement that, we should also think about how an import may work if something inside our databases will change. Even the wordings might change then, as we won't export or import our databases anymore. We will "dump" and "migrate" them then ;)

But that's all another story.

**Pull request merged**
